livox_ros2_driver/livox_ros2_driver/lddc.cpp

```diff
@@ -712,433 +702,349 @@ void Lddc::initializeDiagnostics()
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_internal_voltage-{}", broadcast_code),
+               std::bind(&Lddc::checkVoltage, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_motor_status-{}", broadcast_code),
+               std::bind(&Lddc::checkMotor, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_optical_window-{}", broadcast_code),
+               std::bind(&Lddc::checkDirty, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_firmware_status-{}", broadcast_code),
+               std::bind(&Lddc::checkFirmware, this, std::placeholders::_1,
+                         broadcast_code));
+
+  if (check_pps_signal_) {
+    updater_.add(fmt::format("livox_pps_signal-{}", broadcast_code),
+                 std::bind(&Lddc::checkPPSSignal, this, std::placeholders::_1,
+                           broadcast_code));
+  }
+
+  updater_.add(fmt::format("livox_service_life-{}", broadcast_code),
+               std::bind(&Lddc::checkServiceLife, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(
+      fmt::format("livox_fan_status-{}", broadcast_code),
+      std::bind(&Lddc::checkFan, this, std::placeholders::_1, broadcast_code));
+
+  updater_.add(fmt::format("livox_time_sync-{}", broadcast_code),
+               std::bind(&Lddc::checkTimeSync, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_connection-{}", broadcast_code),
+               std::bind(&Lddc::checkConnection, this, std::placeholders::_1,
+                         broadcast_code));
+}
+
 void Lddc::onDiagnosticsTimer()
 {
   lidar_count_ = 0;
   for (uint32_t i = 0; i < kMaxSourceLidar; ++i) {
     if (lds_->lidars_[i].handle != kMaxSourceLidar) ++lidar_count_;
   }
-  updater_.force_update();
-}
 
-void Lddc::checkTemperature(diagnostic_updater::DiagnosticStatusWrapper & stat)
-{
-  if (lidar_count_ == 0) {
-    stat.summary(DiagStatus::WARN, "No LiDARs Connected");
-    return;
+  for (const auto &lidar : lds_->connected_lidars_) {
+    const std::string broadcast_code = lidar.first;
+    if (registered_code_set_.count(broadcast_code) == 0) {
+      Lddc::registerDiagnosticsUpdater(broadcast_code);
+      registered_code_set_.insert(broadcast_code);
+    }
   }
 
-  int whole_level = DiagStatus::OK;
-
-  for (const auto & lidar : lds_->connected_lidars_) {
-    int level = DiagStatus::OK;
-
-    const auto & broadcast_code = lidar.first;
-    const auto & device_info = lidar.second;
+  updater_.force_update();
+}
 
-    if (device_info == nullptr) {
-      stat.add(broadcast_code, "disconnected");
-      continue;
+boost::optional<DeviceInfo*> Lddc::findDeviceInfoByBroadcastcode(const std::string &broadcast_code) {
```

Comment:
Can you return `boost::optional<DeviceInfo>` here instead of pointer?
I believe you can because DeviceInfo is just a data struct.

Then your code will be simple.

before
```
(*device_info)->status.status_code.lidar_error_code.temp_status);
```

after
```
device_info->status.status_code.lidar_error_code.temp_status);
```

Reply:
If you don't change the type of `connected_lidars_`, the code will be like this.

```cpp
    if (connected_lidar.first == broadcast_code) {
      if (connected_lidar.second == nullptr) {
        return {};
      } 
      return *connected_lidar.second;
    }
```

Or you can assume it's not `nullptr`, then

```cpp
    if (connected_lidar.first == broadcast_code) {
      return *connected_lidar.second;
    }
```

Reply:
What is returned if `broadcast_code` is not found when code is
```
    if (connected_lidar.first == broadcast_code) {
      return *connected_lidar.second;
    }
```

You meant
```
  for (const auto &connected_lidar : lds_->connected_lidars_) {
    if (connected_lidar.first == broadcast_code) {
      return *connected_lidar.second;
    }
  }
  return {};
```

Reply:
Discussion results:
- Change `connected_lidar_` type to non-pointer
    - Remove record when disconnected
- Change `findDeviceInfoByBroadcastcode` type to non-pointer

Reply:
Thanks for your advice. I force-pushed.